hector_math/include/hector_math/containers/ring_buffer.h

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2021 Aljoscha Schmidt, Stefan Fabian. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef HECTOR_MATH_RING_BUFFER_H
+#define HECTOR_MATH_RING_BUFFER_H
+
+#include <array>
+#include <assert.h>
+#include <type_traits>
+
+namespace hector_math
+{
+
+template<typename T, int ElementSize>
+class RingBuffer
+{
+public:
+  // template parameter controls whether the iterator returns const references
+  template<bool IS_CONST>
+  struct ring_iterator {
+    using iterator_category = std::random_access_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    // using difference_type = typename std::iterator<std::random_access_iterator_tag, T>::difference_type;
+    using value_type = T;
+    using pointer = std::conditional_t<IS_CONST, const T *, T *>;
+    using reference = std::conditional_t<IS_CONST, const T &, T &>;
+    using const_reference = T const &;
+
+    ring_iterator( RingBuffer<T, ElementSize> *buffer, size_t index )
+        : buffer_( buffer ), index_( index )
+
+    {
+    }
+
+    template<bool Z = IS_CONST, typename std::enable_if<( !Z ), int>::type * = nullptr>
```

Comment:
This is probably not necessary anymore since reference is now const if `IS_CONST` is true.